homeassistant/helpers/service.py

```diff
@@ -783,3 +783,43 @@ async def check_permissions(call: ServiceCall) -> Any:
         return check_permissions
 
     return decorator
+
+
+class ReloadServiceHelper:
+    """Helper for reload services to minimize unnecessary reloads."""
+
+    def __init__(self, service_func: Callable[[ServiceCall], Awaitable]):
+        """Initialize ReloadServiceHelper."""
+        self._service_func = service_func
+        self._service_running = False
+        self._service_condition = asyncio.Condition()
+
+    async def execute_service(self, service_call: ServiceCall) -> None:
+        """Execute the service.
+
+        If a previous reload task if currently in progress, wait for it to finish first.
+        Once the previous reload task has finished, one of the waiting tasks will be
+        assigned to execute the reload, the others will wait for the reload to finish.
+        """
+
+        do_reload = False
+        async with self._service_condition:
```

Comment:
Why acquire the lock here and then release it? You're going to get racing again. Why not check if `self._service_condition.locked()` ?

Reply:
The lock is there to ensure exclusive access to `self._service_running`